lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,472 @@
-                continue;
-              size_t inPortBaseId;
-              if (!getStorageLocIdFromVal(instPort.value(), inPortBaseId))
-                continue;
-              if (referencedModule) {
-                auto port = referencedModule.getBodyBlock()->getArgument(
-                    instPort.index());
-                // Get all the ports that have a comb path from `port`.
-                for (const auto &portNode : portPaths[port]) {
-                  for (auto outputPort : portNode.getSecond()) {
-                    auto instResult = ins.getResult(
-                        outputPort.val.cast<BlockArgument>().getArgNumber());
-                    size_t outPortId;
-                    if (!getStorageLocIdFromVal(instResult, outPortId))
-                      continue;
-                    Value outLeafVal;
-                    if (!getValueFromStorageLocId(
-                            outPortId + outputPort.fieldId, outLeafVal))
-                      continue;
-                    storageLocIdToValuesMap[inPortBaseId + portNode.first]
-                        .push_back(outLeafVal);
-                  }
-                }
-              }
-            }
+          // Wire is added to the worklist
+          .Case<WireOp>([&](WireOp wire) {
+            worklist.push_back(wire.getResult());
+            if (!wire.getType().cast<FIRRTLBaseType>().isGround())
+              setValRefsTo(wire.getResult(), Node(wire.getResult(), 0));
           })
-          .Case<MemOp>([&](MemOp mem) {
-            if (!(mem.getReadLatency() == 0)) {
-              return;
-            }
-            for (auto memPort : mem.getResults()) {
-              if (!memPort.getType().isa<FIRRTLBaseType>())
-                continue;
+          // All sub elements are added to the worklist.
+          .Case<SubfieldOp>([&](SubfieldOp sub) {
+            auto res = sub.getResult();
+            bool isValid = false;
+            auto fieldIndex = sub.getFieldIndex() + 1;
+            if (memPorts.contains(sub.getInput())) {
+              auto memPort = sub.getInput();
               auto type = memPort.getType().cast<BundleType>();
               auto enableFieldId =
                   type.getFieldID((unsigned)ReadPortSubfield::en);
               auto dataFieldId =
                   type.getFieldID((unsigned)ReadPortSubfield::data);
               auto addressFieldId =
                   type.getFieldID((unsigned)ReadPortSubfield::addr);
-              llvm::SmallDenseSet<size_t> fieldIdsToAdd = {
-                  enableFieldId, dataFieldId, addressFieldId};
-              gatherAggregateLeafs(memPort, rootValues, fieldIdsToAdd);
-
-              size_t baseId;
-              if (!getStorageLocIdFromVal(memPort, baseId))
-                continue;
-              Value dataField;
-              if (!getValueFromStorageLocId(baseId + dataFieldId, dataField))
+              memPorts.insert(memPort);
```

Comment:
Line 252 ensures that `memPort` is already inserted.

Reply:
Thanks for spotting this.